### PR TITLE
[Kernel] [#2252] Use a SnapshotHint to bound protocol and metadata loading when loading latest Snapshot

### LINKED ISSUE #2252: [Kernel] [Feature Request] Incremental snapshot protocol & metadata loading
#### Which Delta project/connector is this regarding?
Kernel

### Overview

Similar to https://github.com/delta-io/connectors/pull/533.

When a kernel snapshot is initialized, we take the LogSegments (JSON files, checkpoint files) of the table, and parse them in reverse order until we find the latest protocol and metadata.

Suppose we have already loaded snapshot `S1` and version 14, and then we want to update to the latest state of the Delta table, say Snapshot `S2` at version 19. Also suppose there is a checkpoint at version 10.

Well, when we loaded `S1`, we read all jsons at versions (in order) 14, 13, 12, 11, and then we read the checkpoint at version 10. 

When we are parsing the delta log files `S2`, we don't need to read _all_ the json and checkpoint files. We only need to read 19, 18, 17, 16, 15.  At that point, if we haven't found the latest metadata and protocol, then we already know what it is as of version 14 since we have S1 in memory already!

### Motivation

Performance boost.

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [X] No. I cannot contribute this feature at this time.

### PR DESCRIPTION
**Pull request by scottsand-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

Resolves #2252

TLDR: this PR increase the performance of multiple `getLatestSnapshot` calls against the same `Table` instance by an average of 8x, and on most commits 34x. It stores some small state in memory, and bounds the computation of future Snapshots.

This PR creates a `SnapshotHint` class. `Snapshot` instances, whenever they decide to load their latest protocol and metadata, will create a new hint and register it with the `SnapshotManager`. `SnapshotManager` will then pass _the latest hint_ to future `Snapshot`s it loads (via the `getLatestSnapshot` API).

When future `Snapshot`s then go to load their latest protocol and metadata (suppose they are loading version 17, and the latest checkpoint is at version 10, and the hint is at version 15), then they can bound their computation and JSON reading/loading, since they need only read JSONs 17 and 16. After they, they can use the hint to know the latest protocol and metadata as of version 15.

## How was this patch tested?

New test suite, `LogReplayMetricsSuite` that tests these edge cases quite well.

Also, tested the performance locally myself:

I tested this against a table with 1M+ add files. I tested master first, and then tested my PR, which was strictly reading a (slightly) larger table.

The test setup was:
- get the latest snapshot with kernel. get the schema and version. record how long it takes to do this
- write to the table with spark
- repeat 20 times.

Master results (milliseconds):
- (747, 292, 294, 282, 279, 287, 278, 269, 279, 278, 272, 278, 279, 275, 279, 262, 273, 284, 281, 267, 277)
- average after 1st commit: 278

My PR (milliseconds):
- (783, 8, 7, 7, 8, 7, 10, 7, 287, 9, 7, 7, 7, 8, 7, 8, 7, 8, 286, 10, 8)
- average after 1st commit: 35, median is 9
- why does it spike up to 287ms? on every 10th commit we read, we are reading the parquet file. we can optimize this away to read the JSON file and then use the hint.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 1]
```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
+                            validateSupportedTable(protocol, metadata);
+                            return new Tuple2<>(protocol, metadata);
+                        }
+                    } else {
+                        final ColumnVector metadataVector = columnarBatch.getColumnVector(1);
 
-                    for (int i = 0; i < metadataVector.getSize(); i++) {
-                        if (!metadataVector.isNullAt(i)) {
-                            metadata = Metadata.fromColumnVector(metadataVector, i, tableClient);
+                        for (int i = 0; i < metadataVector.getSize(); i++) {
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:204]: just indentation

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 2]
```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
+                        // short circuit and use the hint's protocol.
+                        protocol = snapshotHint.get().getProtocol();
+
+                        if (metadata != null) {
+                            // Stop since we have found the latest Protocol and Metadata.
+                            validateSupportedTable(protocol, metadata);
+                            return new Tuple2<>(protocol, metadata);
+                        }
+                    } else {
+                        final ColumnVector protocolVector = columnarBatch.getColumnVector(0);
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:172]: just indentation change

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java [hunk 3]
```diff
@@ -47,14 +47,16 @@ public static Table forPath(TableClient tableClient, String path)
 
     private final Path logPath;
     private final Path dataPath;
+    private final SnapshotManager snapshotManager;
 
     public TableImpl(Path logPath, Path dataPath) {
         this.logPath = logPath;
         this.dataPath = dataPath;
+        this.snapshotManager = new SnapshotManager();
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: @vkorukanti we should have a design philosophy for when and how we construct objects.
what if this constructor eventually gets expensive? etc.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 4]
```diff
@@ -36,23 +40,34 @@ public class SnapshotImpl implements Snapshot {
...
             logPath,
             dataPath,
             tableClient,
-            logSegment);
-        this.protocolAndMetadata = new Lazy<>(logReplay::loadProtocolAndMetadata);
+            logSegment,
+            snapshotHint);
+        this.protocolAndMetadata = new Lazy<>(() -> {
+            // Construct the SnapshotHint lazily. i.e. only create it when some caller/consumer of
+            // this SnapshotImpl instance decides to load the protocol and metadata.
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:64]: In Delta spark, the metadata and protocol are needed ~immediately (**) and so cannot really be lazy. Is that not the case here?

(**) i.e. query planner triggered the snapshot because it needed the table schema

Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:64]: https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala#L169

We are not doing this yet. Not sure why. @vkorukanti ?

Comment (vkorukanti) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:64]: We [validate](https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java#L179) as part of the loading the protocol and metadata. 

If I remember correctly, we want the cost of asking for a snapshot should be quick and not load something unless until is asked. In the case of the query planner, it will be asking for `Snapshot.getSchema` so that should trigger the loading.

We can change to follow the Delta-Spark model.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: It is general practice to not make the constructor expensive, what is special about Kernel?

Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: One this with this change is, we have a state in `Table`. We need to worry about thread safety. For example one thread is loading the `Table.getLatestSnapshot` and other thread calling `Table.getSnapshotAtVersion`. 

We may need to mark the APIs as thread-safe or make sure there are no concurrency issues.

Comment (vkorukanti) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java#L179

**Review by vkorukanti — COMMENTED** [review]
lgtm.

Comment about thread-safety on APIs and tests.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 5]
```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
+                        // Our snapshot hint already tells us the latest Protocol at that hint's
+                        // table version. If we haven't yet found a newer protocol, then we can
+                        // short circuit and use the hint's protocol.
+                        protocol = snapshotHint.get().getProtocol();
+
+                        if (metadata != null) {
+                            // Stop since we have found the latest Protocol and Metadata.
+                            validateSupportedTable(protocol, metadata);
+                            return new Tuple2<>(protocol, metadata);
+                        }
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:170]: This is redundant with the protocol code below. Unlike the column batches, we _know_ a hint _always_ provides both P&M. So we can check the whole thing once (and the hint might even just return the tuple directly, since we no longer look at the components individually).
```scala
while (reverseIterator.hasNext()) {
    ...
  if (snapshotHint.isPresent() && version <= snapshotHint.get().getVersion()) {
    return snapshotHint.get().pAndM()
  }
  if (protocol == null) {
    ...
  }
  if (metadata == null) {
    ...
  }
}
```
If we could safely assume that the hint is always for a version earlier than the log replay version, we could potentially avoid loading the column batch for the hinted version:
```scala
while (reverseIterator.hasNext()) {
    ...
  if (protocol == null) {
    ...
  }
  if (metadata == null) {
    ...
  }
  // Before starting the next iteration, check if the hint satisfies it.
  if (snapshotHint.isPresent() && version-1 == snapshotHint.get().getVersion()) {
    return snapshotHint.get().pAndM()
  }
}
```
If we had a way to lower-bound filter `allLogFilesReversed` by the hint version, then it would be even more efficient (no extra I/O), but more complicated to set up:
```scala
final Iterator<FileStatus> selectedLogFilesReversed = ...
try(...) {
  while (reverseIterator.hasNext()) {
    if (protocol == null) {
      ...
    }
    if (metadata == null) {
      ...
    }
  }
} catch (...) { ...}

if (snapshotHint.isPresent() && version <= snapshotHint.get().getVersion()) {
  return snapshotHint.get().pAndM()
}

if (protocol == null) {
```

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 6]
```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
-                final ColumnarBatch columnarBatch = reverseIter.next()._1.getData();
+                final ActionIterElem nextElem = reverseIter.next();
+                final ColumnarBatch columnarBatch = nextElem.getFileDataReadResult().getData();
+                final long version = nextElem.getVersion();
 
                 assert(columnarBatch.getSchema().equals(PROTOCOL_METADATA_READ_SCHEMA));
 
                 if (protocol == null) {
-                    final ColumnVector protocolVector = columnarBatch.getColumnVector(0);
+                    if (snapshotHint.isPresent() && version <= snapshotHint.get().getVersion()) {
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:160]: This should be `==` -- `<=` isn't safe in the general case where the hint could be earlier than the earliest log segment version (e.g. hint at v101, checkpoint at v105). Granted, in such cases we _should_ always get a P&M from the checkpoint, but if somehow the checkpoint were incorrect/incomplete, I don't think we'd want the hint to hide that fact?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: What is the thread unsafe concern here? What's unsafe about calling `.getLatestSnapshot` and then calling `.getSnapshotAtVersion`?

When these snapshots register themselves as the latest snapshot hint with the SnapshotManager, we check their version.

BTW, would be great if we wanted to keep a fixed size-bounded cache of snapshot hints , eh?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 7]
```diff
@@ -136,51 +139,81 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
+        try (CloseableIterator<ActionIterElem> reverseIter =
                  new ActionsIterator(
                      tableClient,
                      logSegment.allLogFilesReversed(),
                      PROTOCOL_METADATA_READ_SCHEMA)) {
             while (reverseIter.hasNext()) {
-                final ColumnarBatch columnarBatch = reverseIter.next()._1.getData();
+                final ActionIterElem nextElem = reverseIter.next();
+                final ColumnarBatch columnarBatch = nextElem.getFileDataReadResult().getData();
+                final long version = nextElem.getVersion();
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:155]: > then we just need to defer calling it until after we've done our version checks

Nice catch! SGTM!

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 8]
```diff
@@ -95,20 +96,22 @@ public class LogReplay implements Logging {
...
+            LogSegment logSegment,
+            Optional<SnapshotHint> snapshotHint) {
         assertLogFilesBelongToTable(logPath, logSegment.allLogFilesUnsorted());
 
         this.dataPath = dataPath;
         this.logSegment = logSegment;
         this.tableClient = tableClient;
-        this.protocolAndMetadata = new Lazy<>(this::loadTableProtocolAndMetadata);
+        this.protocolAndMetadata = new Lazy<>(
+            () -> this.loadTableProtocolAndMetadata(snapshotHint)
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:113]: Yup. It's all about API contracts. Do we want `loadTableProtocolAndMetadata` to return a `validated` P & M?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 5 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:170]: Just remember that we may find a Protocol but NOT find the Metadata (and vice versa). Thus, just because we get to the hint's version, we cannot just return the hint's protocol AND metadata. We have to do null checks to see if we found an earlier one already.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 6 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:160]: Agreed!

Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:160]: Agreed it should be `==` ... but just FYI in your example, `snapshotHint.get().getVersion() = 101`, `version=105`

then `version <= snapshotHint.get().getVersion()` evaluates to `false` so we wouldn't use the hint

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: Sorry, pardon?

**Comment by scottsand-db** [comment]
> lgtm.
> 
> Comment about thread-safety on APIs and tests.

@vkorukanti - I don't see your tests comment. Right now, we expose no metrics to verify that we read fewer files. We also can't write ... so we also can't meaningfully update our delta table to get a newer snapshot? Unless we copy files from goldenTable1 to goldenTable2 etc ... that is messy. Can we just use spark to test? What do you think?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 9]
```diff
@@ -199,6 +210,10 @@ private Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
                             return new Tuple2<>(protocol, metadata);
                         }
                     } else {
+                        if (columnarBatch == null) {
+                            columnarBatch = nextElem.getFileDataReadResult().getData();
+                            assert(columnarBatch.getSchema().equals(PROTOCOL_METADATA_READ_SCHEMA));
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:215]: Why do we need to assert the schema? Is there some scenario where it would not match?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java [hunk 10]
```diff
@@ -90,6 +95,21 @@ public Snapshot buildLatestSnapshot(TableClient tableClient, Path logPath, Path
...
+     * newer (i.e. has a later table version).
+     * <p>
+     * This doesn't need to be thread-safe. Suppose the current `latestSnapshotHint` is at version
+     * N. If three Snapshots (N-1, N+1, N+2) concurrently register their hints, only N+1 and N+2
+     * will be able to replace the current. And either winning the update is fine.
+     */
+    public void registerHint(SnapshotHint newHint) {
+        if (!latestSnapshotHint.isPresent() ||
+                newHint.getVersion() > latestSnapshotHint.get().getVersion()) {
+            latestSnapshotHint = Optional.of(newHint);
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java:109]: This code is not thread safe, in spite of the comment stating otherwise. It reads `latestSnapshotHint` twice, and there's no guarantee it stays the same between reads. And then it updates the hint as well, and there's nothing to block the hint from getting changed by somebody else who races in (which could break monotonicity of the hint). Further, the value is not marked volatile, so there's no guarantee that different threads _ever_ see each other's changes.

Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java:109]: Should probably use an [AtomicReference](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReference.html#get--) instead of `Optional`.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java (hunk 4 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:64]: We already had to perform I/O to create the snapshot at all (LIST) so I don't see much benefit in being lazy about P&M. Especially given the complexity and code tangling it introduces.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: > What is the thread unsafe concern here?
See my other comment about this.

> would be great if we wanted to keep a fixed size-bounded cache of snapshot hints , eh?
At some point we have definitely wandered into the anti-goal of complexity in the name of optimizing performance. 

The actual goal  is to be as simple as possible while not pessimizing performance. If we really want to allow optimization here, we should probably allow the caller to pass in the hint directly (from whatever snapshot they like), rather than taking an opinion about how many snapshots/hints to remember internally.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 11]
```diff
@@ -36,23 +40,34 @@ public class SnapshotImpl implements Snapshot {
...
+            final Tuple2<Protocol, Metadata> pAndM = logReplay.loadProtocolAndMetadata();
+            final SnapshotHint hint = new SnapshotHint(version, pAndM._1, pAndM._2);
+            snapshotManager.registerHint(hint);
+            return pAndM;
+        });
+        this.snapshotHint = snapshotHint;
```
Comment (ryan-johnson-databricks) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:70]: Why does this need to be a member? The value gets used by the lazy computation, but this field has no use sites?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: Good catch... we've been intentionally avoiding internal state like this so far.

And it's especially messy because the snapshot manager is captured by the snapshot as well, but only inside a lazy value that will make it ~impossible to guess lifetimes.

Following the other comment -- if we accept that creating a snapshot _will_ access the table at least enough to compute P&M, then we can get rid of at least some of this messiness by having `SnapshotManager.createSnapshot` set the hint directly:
```java
private SnapshotImpl createSnapshot(...) {
    ...
  final val snapshot = new SnapshotImpl(...);
  final SnapshotHint hint = new SnapshotHint(
    snapshot.getVersion(tableClient), snapshot.getProtocol(), snapshot.getMetadata());
  registerHint(hint);
  return snapshot;
}
```
(something like the above would get rid of two lazy values and remove the need to pass the snapshot manager into the snapshot constructor)

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Given that the method is only used to assign a private member, there isn't an API contract yet. We just have to validate before exposing it to the outside world, and this is a narrow waist, no?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 5 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Right... I had it backward. The only way the `==` vs `<=` matters is if we had a gap in the version number sequence, but the listing logic already checked for that.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 6 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Right... I had it backward. The only way the `==` vs `<=` matters is if we had a gap in the version number sequence, but the listing logic already checked for that.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 5 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Good point.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: What is you want to construct a LogReplay but not get the protocol and metadata?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 9 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:215]: This is part of a bigger topic. We have this opage `ColumnarBatch columnarBatch` used all throughout kernel.

Say I have a method, `foo`, that takes as input some columnarBatch. What is it's schem? As a developer, how do I write code that uses `foo`? I need to know what columns it has. How do we check this? How do we define this? 

`assert` is one way to do it. do you prefer something else? or nothing at all?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: I'd prefer to see a compelling example of that use case before optimizing for it. But LogReplay is anyway internal, so it's really "what if you want to construct a Snapshot but not get P&M?"

We don't want to eagerly grab app ids because there's no guarantee the user will access it.
We don't eagerly replay file actions because pruning will hopefully eliminate most of them.

But it's impossible to interact meaningfully with a Delta table without knowing P&M -- all you can learn without them is the latest commit version, and we can get that from the log segment without constructing a LogReplay at all.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java (hunk 9 above)
Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:215]: Yeah, you're right that assert makes sense. It's not a user error when this happens, but rather a connector bug. In _theory_ the connector should have honored kernel's request for a read with the given schema, but in practice a connector bug would lead to arbitrary "fun" if we didn't notice early.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:55]: I will make all this not `lazy` in a followup pr

Comment (ryan-johnson-databricks) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java:54]: Can we please get rid of the lazy now, not in a future PR? It's weird to be changing the API and adding complexity, just to accommodate something should be removed.

**Review by ryan-johnson-databricks — COMMENTED** [review]
Looking good, just one remaining significant feedback, about how the lazy code complicates things...

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java [hunk 12]
```diff
@@ -136,51 +141,94 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches() {
...
+                final ActionIterElem nextElem = reverseIter.next();
+                final long version = nextElem.getVersion();
 
-                assert(columnarBatch.getSchema().equals(PROTOCOL_METADATA_READ_SCHEMA));
+                // Load this lazily (as needed). We may be able to just use the hint.
+                ColumnarBatch columnarBatch = null;
 
                 if (protocol == null) {
-                    final ColumnVector protocolVector = columnarBatch.getColumnVector(0);
+                    if (snapshotHint.isPresent() && version == snapshotHint.get().getVersion()) {
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Thanks. I've simplified it.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionIterElem.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.data.FileDataReadResult;
+
+class ActionIterElem {
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionIterElem.java:22]: docs? better name? DeltaLogActions?

**Review by vkorukanti — APPROVED** [review]
